dockerfiles/ruby/solargraph.go

```diff
@@ -0,0 +1,207 @@
+package main
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/url"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/sourcegraph/jsonrpc2"
+)
+
+// lazyObjectStream wraps a jsonrpc2.ObjectStream to only start the connection
+// once we have received an initialize request.
+type lazyObjectStream struct {
+	Connect func(rootURI string) (jsonrpc2.ObjectStream, error)
+
+	once   *sync.Once    // protects writes to stream, err and closing ready.
+	ready  chan struct{} // protects reads to stream and err.
+	stream jsonrpc2.ObjectStream
+	err    error
+}
+
+func (t *lazyObjectStream) WriteObject(v interface{}) error {
+	t.once.Do(func() {
+		defer close(t.ready)
+
+		// minimal json to get at rootUri
+		var msg struct {
+			Method string `json:"method"`
+			Params struct {
+				RootURI string `json:"rootUri"`
+			} `json:"params"`
+		}
+		b, err := json.Marshal(v)
+		if err != nil {
+			t.err = err
+			return
+		}
+		if err = json.Unmarshal(b, &msg); err != nil {
+			t.err = err
+			return
+		}
+
+		if msg.Method != "initialize" {
+			t.err = fmt.Errorf("expected first message to be initialize, got %s", msg.Method)
+			return
+		}
+
+		t.stream, t.err = t.Connect(msg.Params.RootURI)
+	})
+	if t.err != nil {
+		return t.err
+	}
+	return t.stream.WriteObject(v)
+}
+
+func (t *lazyObjectStream) ReadObject(v interface{}) error {
+	// we wait for the initialize request (via WriteObject) to happen first.
+	<-t.ready
+	if t.err != nil {
+		return t.err
+	}
+	return t.stream.ReadObject(v)
+}
+
+func (t *lazyObjectStream) Close() error {
+	t.once.Do(func() {
+		t.err = jsonrpc2.ErrClosed
+		close(t.ready)
+	})
+	if t.err != nil {
+		return nil
+	}
+	return t.stream.Close()
+}
+
+func startSolargraph(ctx context.Context, dir string) (*exec.Cmd, int, error) {
+	var (
+		stderr io.Reader
+		err    error
+	)
+	cmd := exec.CommandContext(ctx, "solargraph", "socket", "--port", "0")
+	cmd.Dir = dir // set CWD env var?
+	cmd.Stdout = os.Stderr
+	stderr, err = cmd.StderrPipe()
+	if err != nil {
+		return nil, 0, err
+	}
+	stderr = io.TeeReader(stderr, os.Stderr)
+	if err := cmd.Start(); err != nil {
+		return nil, 0, err
+	}
+
+	scanner := bufio.NewScanner(stderr)
+	scanner.Split(bufio.ScanWords)
+	for scanner.Scan() {
+		word := scanner.Text()
+		if strings.HasPrefix(word, "PORT=") {
+			// read the rest of stderr to keep the tee reader going
+			go io.Copy(ioutil.Discard, stderr)
+
+			port, err := strconv.Atoi(word[len("PORT="):])
+			if err != nil {
+				break
+			}
+			return cmd, port, nil
+		}
+	}
+	// if we get to this point we didn't find the port. cleanup
```

Comment:
When will `Scan()` terminate? When the process exits?

Reply:
yup. Basically when `stderr` returns an error (normally `io.EOF`). Hopefully before then we see the word `PORT=`, in which case we terminate the scanner earlier. This is essentially what solargraph-util does. Note we ensure we always fully read `stderr` to prevent blocking the process.